[dask] add tutorial documentation (fixes #3814, fixes #3838) (#4030)
microsoft/LightGBM: docs/Parallel-Learning-Guide.rst

LightGBM's Python package supports distributed learning via `Dask`_. This integration is maintained by LightGBM's maintainers.

Dask Examples
'''''''''''''

For sample code using ``lightgbm.dask``, see `these Dask examples`_ .

Training with Dask
''''''''''''''''''

This section contains detailed information on performing LightGBM distributed training using Dask.

Configuring the Dask Cluster
****************************

**Allocating Threads**

When setting up a Dask cluster for training, give each Dask worker process at least two threads. If you do not do this, training might be substantially slower because communication work and training work will block each other.

If you do not have other significant processes competing with Dask for resources, just accept the default ``nthreads`` from your chosen ``dask.distributed`` cluster.

.. code:: python

  from distributed import Client, LocalCluster

  cluster = LocalCluster(n_workers=3)
  client = Client(cluster)

**Managing Memory**

> Use the Dask diagnostic dashboard or your preferred monitoring tool to monitor Dask workers' memory consumption during training. As described in `the Dask worker documentation`_, Dask workers will automatically start spilling data to Disk if memory consumption gets too high. This can substantially slow down computations, since disk I/O is usually much slower than reading the same data from memory.

Author: 😆

LightGBM's Dask estimators support setting an attribute ``client`` to control the client that is used.

.. code:: python

  import lightgbm as lgb
  from distributed import LocalCluster, Client

  cluster = LocalCluster()
  client = Client(cluster)

  # option 1: keyword argument in constructor
  dask_model = lgb.DaskLGBMClassifier(client=client)

  # option 2: set_params() after construction
  dask_model = lgb.DaskLGBMClassifier()
  dask_model.set_params(client=client)

Using Specific Ports
********************

At the beginning of training, ``lightgbm.dask`` sets up a LightGBM network where each Dask worker runs one long-running task that acts as a LightGBM worker. During training, LightGBM workers communicate with each other over TCP sockets. By default, random open ports are used when creating these sockets.

If the communication between Dask workers in the cluster used for training is restricted by firewall rules, you must tell LightGBM exactly what ports to use.

> **Option 1: Provide a specific list of addresses and ports**

Reviewer: ```suggestion
**Option 1: provide a specific list of addresses and ports**
```

  from distributed import Client, LocalCluster

  cluster = LocalCluster(n_workers=2)
  client = Client(cluster)

  X = da.random.random((1000, 10), (500, 10))
  y = da.random.random((1000,), (500,))

  dask_model = lgb.DaskLGBMRegressor()
  dask_model.fit(X, y)

  with open("dask-model.pkl", "wb") as f:
      pickle.dump(dask_model, f)

A model saved this way can then later be loaded with whichever serialization library you used to save it.

.. code:: python

  import pickle
  with open("dask-model.pkl", "rb") as f:
      dask_model = pickle.load(f)

.. note::

  If you explicitly set a Dask client (see `Using a Specific Dask Client <#using-a-specific-dask-client>`__), it will not be saved when pickling the estimator. When loading a Dask estimator from disk, if you need to use a specific client you can add it after loading with ``dask_model.set_params(client=client)``.

**Option 2: pickle the sklearn estimator**

The estimators available from ``lightgbm.dask`` can be converted to an instance of the equivalent class from ``lightgbm.sklearn``. Choosing this option allows you to use Dask for training but avoid depending on any Dask libraries at scoring time.

.. code:: python

>   import dask.array as da
>   import pickle
>   import lightgbm as lgb
>   from distributed import Client, LocalCluster
> 
>   cluster = LocalCluster(n_workers=2)
>   client = Client(cluster)
> 
>   X = da.random.random((1000, 10), (500, 10))
>   y = da.random.random((1000,), (500,))
> 
>   dask_model = lgb.DaskLGBMRegressor()
>   dask_model.fit(X, y)
> 
>   # convert to sklearn equivalent
>   sklearn_model = dask_model.to_local()
> 
>   print(type(sklearn_model))
>   #> lightgbm.sklearn.LGBMRegressor
> 
>   with open("sklearn-model.pkl", "wb") as f:
>       pickle.dump(sklearn_model, f)
> 
> A model saved this way can then later be loaded with whichever serialization library you used to save it.
> 
> .. code:: python
> 
>   import pickle
>   with open("sklearn-model.pkl", "rb") as f:
>       sklearn_model = pickle.load(f)

Reviewer: It is advised to use `joblib` for sklearn estimators.
> In the specific case of scikit-learn, it may be better to use joblib’s replacement of pickle (`dump` & `load`), which is more efficient on objects that carry large numpy arrays internally as is often the case for fitted scikit-learn estimators, but can only pickle to the disk and not to a string:
   https://scikit-learn.org/stable/modules/model_persistence.html#python-specific-serialization

Also, it brings some variety to examples.

```suggestion
  import dask.array as da
  import joblib
  import lightgbm as lgb
  from distributed import Client, LocalCluster

  cluster = LocalCluster(n_workers=2)
  client = Client(cluster)

  X = da.random.random((1000, 10), (500, 10))
  y = da.random.random((1000,), (500,))

  dask_model = lgb.DaskLGBMRegressor()
  dask_model.fit(X, y)

  # convert to sklearn equivalent
  sklearn_model = dask_model.to_local()

  print(type(sklearn_model))
  #> lightgbm.sklearn.LGBMRegressor

  joblib.dump(sklearn_model, "sklearn-model.joblib")

A model saved this way can then later be loaded with whichever serialization library you used to save it.

.. code:: python

  import joblib

  sklearn_model = joblib.load("sklearn-model.joblib")
```

Author: oh sure! I like it

**Option 3: save the LightGBM Booster**

The lowest-level model object in LightGBM is the ``lightgbm.Booster``. After training, you can extract a Booster from the Dask estimator.

.. code:: python

>   import dask.array as da
>   import pickle

Reviewer: Not used in this example.
```suggestion
  import dask.array as da
```

  import lightgbm as lgb
  from distributed import Client, LocalCluster

  cluster = LocalCluster(n_workers=2)
  client = Client(cluster)

  X = da.random.random((1000, 10), (500, 10))
  y = da.random.random((1000,), (500,))

  dask_model = lgb.DaskLGBMRegressor()
  dask_model.fit(X, y)

>   # convert to sklearn equivalent

Reviewer: ```suggestion
  # get underlying Booster object
```

  bst = dask_model.booster_

> From the point forward, you can use any of the following methods to save the Booster.

Reviewer: ```suggestion
From the point forward, you can use any of the following methods to save the Booster:
```

LightGBM's Python package supports distributed learning via `Dask`_. This integration is maintained by LightGBM's maintainers.

Quick Dask Examples
'''''''''''''''''''

Theses examples show minimal code needed to train a LightGBM model using Dask. See the other sections after this for more details.

**Classification**

For multiclass classification, set ``centers=3`` below.

> .. code:: python

Reviewer: @jameslamb I see this PR is WIP but I'd like to suggest one thing that can change the whole design of the PR.

WDYT about writing examples (even "quick") in a form of separate files in the [`examples/python-guide/Dask`](https://github.com/microsoft/LightGBM/tree/master/examples/python-guide) folder and have a reference for them in this doc? Going this way we will run all examples at our CI and be sure they are up-to-date and works fine.
https://github.com/microsoft/LightGBM/blob/87c37bf04f12cdf15498adfec6bf0baba5498ab2/.ci/test.sh#L221

Author: oh sure! I like that idea
